main.cpp

```diff
@@ -294,256 +242,193 @@ struct IntType;
 private:
-    double* value = nullptr;
-    DoubleType& powInternal(double arg);
+    std::unique_ptr<Type> value;
+    Numeric& powInternal(Type arg)
+    {
+        *value = static_cast<Type>(std::pow( *value, arg ));
+        return *this;
+    }
 };
 
-DoubleType& DoubleType::apply(std::function<DoubleType&(double&)> dtLamb)
+template<typename NumericType>
+void myNumericFreeFunct(std::unique_ptr<NumericType>& ftPtr)
 {
-    if ( dtLamb )
-        return dtLamb(*value);
-    return *this;
+    NumericType& ftFreeFunc = *ftPtr;
+    ftFreeFunc += static_cast<NumericType>(7.0);
 }
 
-DoubleType& DoubleType::apply(void(*dtPtr)(double&))
+template<>
+struct Numeric<double>
 {
-    if( dtPtr )
-        dtPtr(*value);
-    return *this;
-}
+    using Type = double;
 
-DoubleType& DoubleType::pow(double rhs)
-{
-    return powInternal(rhs);
-}
+    explicit Numeric(Type ft) : value( std::make_unique<Type>(ft) ) { }
 
-DoubleType& DoubleType::powInternal(double arg)
-{
-    *value = std::pow( *value, arg );
-    return *this;
-}
+    Numeric& operator+=(const Type& rhs)
+    {
+        *value += rhs;
+        return *this;
+    }
 
-DoubleType& DoubleType::operator+=( double rhs )
-{
-    *value += rhs;
-    return *this;
-}
-DoubleType& DoubleType::operator-=( double rhs )
-{
-    *value -= rhs;
-    return *this;
-}
-DoubleType& DoubleType::operator*=( double rhs )
-{
-    *value *= rhs;
-    return *this;
-}
-DoubleType& DoubleType::operator/=( double rhs )
-{
-    if (rhs == 0.0)
-        std::cout << "warning: floating point division by zero!" << "\n";
-        
-    *value /= rhs;
-    return *this;
-}
+    Numeric& operator-=(const Type& rhs)
+    {
+        *value -= rhs;
+        return *this;
+    }
 
-void myDoubleFreeFunct(double& dtPtr)
-{
-    double& dtFreeFunc = dtPtr;
-    dtFreeFunc += 6.0;
-}
+    Numeric& operator*=(const Type& rhs)
+    {
+        *value *= rhs;
+        return *this;
+    }
 
-struct IntType
-{
-    explicit IntType(int it) : value(new int(it)) {}
-    ~IntType()
+    template<typename DividedByType>
+    Numeric& operator/=(const DividedByType& rhs)
     {
-        delete value;
-        value = nullptr;
+        if constexpr (std::is_same<Type, int>::value)
+        {
+            if constexpr (std::is_same<DividedByType, int>::value)
+            {
+                if (rhs == 0)
+                {
+                    std::cout << "error: integer division by zero is an error and will crash the program!\n";
+                    return *this;
+                }
+            }
+            else if (std::abs(rhs) <= std::numeric_limits<DividedByType>::epsilon())
+            {
+                std::cout << "warning: floating point division by zero!" << "\n";
+            }
+        }
+        else if (std::abs(rhs) <= std::numeric_limits<DividedByType>::epsilon())
+        {
+            std::cout << "warning: floating point division by zero!" << "\n";
+        }
+
+        *value /= rhs;
+        return *this;
```

Comment:
`Numeric<double>` doesn't need all of this template checking, because the argument will be automatically converted to a double when you use `const Type&` as the argument type. 
That means you only need to check if `rhs` is `0.0` and respond accordingly.
- [ ] revise